### ISSUE TITLE
init fails to analyze overlay with no functions

### ISSUE THREAD

**Issue by LagoLunatic** [AetiasHax/ds-decomp#1]
Tested on Castlevania: Order of Ecclesia:
```
>dsd init --extract-path ./orig/YR9E00 --output-path ./config/YR9E00 --dry true
[2024-09-09T20:14:22Z INFO  ds_rom::rom::rom] Loading ROM from path ./orig/YR9E00
[2024-09-09T20:14:22Z INFO  ds_rom::rom::rom] Loading ROM assets
[2024-09-09T20:14:22Z WARN  dsd::config::module] Expected .text to end (0x20163e0) where .init starts (0x20d7ff4)
thread 'main' panicked at src\config\module.rs:223:49:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

In this game, overlay 0 doesn't have any code in it, instead it has the game's entire script in it. find_functions currently seems to assume every overlay will have functions in it, so it fails when an empty list is returned:
https://github.com/AetiasHax/ds-decomp/blob/074e28e7ebf104341c0fe329769257bc3b7af5c1/src/config/module.rs#L223-L224

**Comment by AetiasHax** [comment]
> In this game, overlay 0 doesn't have any code in it, instead it has the game's entire script in it.

That's interesting 🤔 Should be a simple fix, but I also noticed this warning:
```
[2024-09-09T20:14:22Z WARN  dsd::config::module] Expected .text to end (0x20163e0) where .init starts (0x20d7ff4)
```

The `.text` section should continue all the way until `.init` starts, unless I'm mistaken about the section order being consistent for all retail games. Regardless, this massive gap is a bigger concern, as the rest of the analysis relies on there being symbols for every function in the game. I'll have to look into the byte contents after `0x20163e0` so I can find out what the function analysis is missing.

**Comment by LagoLunatic** [comment]
Yeah, there are many functions after 0x20163e0 and before 0x20d7ff4 that it misses. 2886 functions, if ghidra's analysis is correct.
The next function after 0x20163e0 starts at 0x020163f4 and it seems to reference a literal pool starting at 0x20163e0.

There are similar warnings for Dawn of Sorrow and Portrait of Ruin (and the same overlay error, because these games also use a few overlays for things like graphics data):
```
>dsd init --extract-path orig/ACVE00 --output-path config/ACVE00 --dry false  
[2024-09-11T20:02:34Z INFO  ds_rom::rom::rom] Loading ROM from path orig/ACVE00
[2024-09-11T20:02:34Z INFO  ds_rom::rom::rom] Loading ROM assets
[2024-09-11T20:02:34Z WARN  dsd::config::module] Expected .text to end (0x20728f4) where .init starts (0x208ac14)
thread 'main' panicked at src\config\module.rs:223:49:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```
```
>dsd init --extract-path orig/ACBE00 --output-path config/ACBE00 --dry false 
[2024-09-11T20:02:48Z INFO  ds_rom::rom::rom] Loading ROM from path orig/ACBE00
[2024-09-11T20:02:48Z INFO  ds_rom::rom::rom] Loading ROM assets
[2024-09-11T20:02:48Z WARN  dsd::config::module] Expected .text to end (0x20a3620) where .init starts (0x20ca560)
thread 'main' panicked at src\config\module.rs:223:49:
called `Option::unwrap()` on a `None` value
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

The PoR one seems similar to the OoE one in that it thinks .text ends at a literal pool that appears before the function that uses it.
The DoS one also ends at a literal pool, but this one appears after the function that uses it, not before.

**Comment by AetiasHax** [comment]
*Castlevania: Order of Ecclesia* is now compatible as of v0.2.1! Or at least as far as section boundary detection goes, they all look good from what I can tell.

I didn't test the other Castlevania games you mentioned, so could you run them through `dsd init` again and see how it goes?

**Comment by LagoLunatic** [comment]
Here is the new output of `init` for each game:

<details open>
<summary><h3>Dawn of Sorrow</h3></summary>
<pre>
>dsd init --rom-config ./orig/ACVE00/config.yaml --output-path ./config/ACVE00 --build-path ./build/ACVE00
[2024-11-12T21:39:44Z INFO  ds_rom::rom::rom] Loading ROM from ./orig/ACVE00/config.yaml
[2024-11-12T21:39:44Z ERROR ds_decomp::analysis::data] Function call from 0x020008c4 in ARM9 main to 0x02075f18 leads to no function
Error: Function call leads nowhere, see above
</pre>
</details>

<details open>
<summary><h3>Portrait of Ruin</h3></summary>
<pre>
>dsd init --rom-config ./orig/ACBE00/config.yaml --output-path ./config/ACBE00 --build-path ./build/ACBE00
[2024-11-12T21:42:24Z INFO  ds_rom::rom::rom] Loading ROM from ./orig/ACBE00/config.yaml
[2024-11-12T21:42:24Z INFO  ds_decomp::analysis::functions] Function at 020a3634 was adjusted to include pre-code constant pool at 020a3620
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] No functions from 0x0209ded4 in ARM9 main to 0x01ffe3a0:
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] No functions from 0x0209def4 in ARM9 main to 0x01ffe3a0:
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] No functions from 0x0209df80 in ARM9 main to 0x01ffe2f4:
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] No functions from 0x0209dfc0 in ARM9 main to 0x01ffe2f4:
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] No functions from 0x0209e034 in ARM9 main to 0x01ffe32c:
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] No functions from 0x0209e0b8 in ARM9 main to 0x01ffe294:
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] No functions from 0x0209e140 in ARM9 main to 0x01ffe32c:
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] No functions from 0x0209e1c4 in ARM9 main to 0x01ffe294:
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] No functions from 0x0209e2d4 in ARM9 main to 0x01ffe3a0:
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] No functions from 0x0209e3d4 in ARM9 main to 0x01ffe3a0:
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] No functions from 0x0209e400 in ARM9 main to 0x01ffe3a0:
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] No functions from 0x0209e9a4 in ARM9 main to 0x01ffe3a0:
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] Function call from 0x020bd744 in ARM9 main to 0x020bdb50 goes to middle of function 'func_020bdb48' at 0x020bdb48, adding an external label symbol
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] Function call from 0x01ff847c in ITCM to 0x01ff833c goes to middle of function 'func_01ff830c' at 0x01ff830c, adding an external label symbol
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] Function call from 0x01ff84a8 in ITCM to 0x01ff833c goes to middle of function 'func_01ff830c' at 0x01ff830c, adding an external label symbol
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] Function call from 0x01ffd508 in ITCM to 0x01ff833c goes to middle of function 'func_01ff830c' at 0x01ff830c, adding an external label symbol
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] Function call from 0x01ffd5e0 in ITCM to 0x01ff833c goes to middle of function 'func_01ff830c' at 0x01ff830c, adding an external label symbol
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] Function call from 0x01ffd630 in ITCM to 0x01ff833c goes to middle of function 'func_01ff830c' at 0x01ff830c, adding an external label symbol
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] Function call from 0x01ffd670 in ITCM to 0x01ff833c goes to middle of function 'func_01ff830c' at 0x01ff830c, adding an external label symbol
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] Function call from 0x01ffd6b0 in ITCM to 0x01ff833c goes to middle of function 'func_01ff830c' at 0x01ff830c, adding an external label symbol
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] Function call from 0x01ffd6f8 in ITCM to 0x01ff833c goes to middle of function 'func_01ff830c' at 0x01ff830c, adding an external label symbol
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] Function call from 0x01ffd730 in ITCM to 0x01ff833c goes to middle of function 'func_01ff830c' at 0x01ff830c, adding an external label symbol
[2024-11-12T21:42:24Z WARN  ds_decomp::analysis::data] Function call from 0x01ffd768 in ITCM to 0x01ff833c goes to middle of function 'func_01ff830c' at 0x01ff830c, adding an external label symbol
</pre>
</details>

<details open>
<summary><h3>Order of Ecclesia</h3></summary>
<pre>
>dsd init --rom-config ./orig/YR9E00/config.yaml --output-path ./config/YR9E00 --build-path ./build/YR9E00 
[2024-11-12T21:42:01Z INFO  ds_rom::rom::rom] Loading ROM from ./orig/YR9E00/config.yaml
[2024-11-12T21:42:01Z INFO  ds_decomp::analysis::functions] Function at 020163f4 was adjusted to include pre-code constant pool at 020163e0
[2024-11-12T21:42:01Z WARN  ds_decomp::analysis::data] Function call from 0x02023e30 in ARM9 main to 0x0202407c goes to middle of function 'func_02024074' at 0x02024074, adding an external label symbol
[2024-11-12T21:42:01Z WARN  ds_decomp::analysis::data] Function call from 0x01ff8cac in ITCM to 0x01ff8b6c goes to middle of function 'func_01ff8b3c' at 0x01ff8b3c, adding an external label symbol
[2024-11-12T21:42:01Z WARN  ds_decomp::analysis::data] Function call from 0x01ff8cd8 in ITCM to 0x01ff8b6c goes to middle of function 'func_01ff8b3c' at 0x01ff8b3c, adding an external label symbol
[2024-11-12T21:42:01Z WARN  ds_decomp::analysis::data] Function call from 0x01ffdd38 in ITCM to 0x01ff8b6c goes to middle of function 'func_01ff8b3c' at 0x01ff8b3c, adding an external label symbol
[2024-11-12T21:42:01Z WARN  ds_decomp::analysis::data] Function call from 0x01ffde10 in ITCM to 0x01ff8b6c goes to middle of function 'func_01ff8b3c' at 0x01ff8b3c, adding an external label symbol
[2024-11-12T21:42:01Z WARN  ds_decomp::analysis::data] Function call from 0x01ffde60 in ITCM to 0x01ff8b6c goes to middle of function 'func_01ff8b3c' at 0x01ff8b3c, adding an external label symbol
[2024-11-12T21:42:01Z WARN  ds_decomp::analysis::data] Function call from 0x01ffdea0 in ITCM to 0x01ff8b6c goes to middle of function 'func_01ff8b3c' at 0x01ff8b3c, adding an external label symbol
[2024-11-12T21:42:01Z WARN  ds_decomp::analysis::data] Function call from 0x01ffdee0 in ITCM to 0x01ff8b6c goes to middle of function 'func_01ff8b3c' at 0x01ff8b3c, adding an external label symbol
[2024-11-12T21:42:01Z WARN  ds_decomp::analysis::data] Function call from 0x01ffdf28 in ITCM to 0x01ff8b6c goes to middle of function 'func_01ff8b3c' at 0x01ff8b3c, adding an external label symbol
[2024-11-12T21:42:01Z WARN  ds_decomp::analysis::data] Function call from 0x01ffdf60 in ITCM to 0x01ff8b6c goes to middle of function 'func_01ff8b3c' at 0x01ff8b3c, adding an external label symbol
[2024-11-12T21:42:01Z WARN  ds_decomp::analysis::data] Function call from 0x01ffdf98 in ITCM to 0x01ff8b6c goes to middle of function 'func_01ff8b3c' at 0x01ff8b3c, adding an external label symbol
[2024-11-12T21:42:01Z WARN  ds_decomp::analysis::data] Function call from 0x01fff0f0 in ITCM to 0x01fff1cc goes to middle of function 'func_01fff0d8' at 0x01fff0d8, adding an external label symbol
[2024-11-12T21:42:01Z WARN  ds_decomp::analysis::data] Function call from 0x01fff0f4 in ITCM to 0x01fff20c goes to middle of function 'func_01fff0d8' at 0x01fff0d8, adding an external label symbol
</pre>
</details>

---

OoE now runs `init` without errors, but it doesn't seem to be fully fixed. When running `delink`:
```
>dsd delink --config-path ./config/YR9E00/arm9/config.yaml 
[2024-11-12T22:11:53Z INFO  ds_rom::rom::rom] Loading ROM from ./config/YR9E00/arm9\../../../orig/YR9E00/config.yaml
Error: function func_020163f4 could not be analyzed: InvalidStart
```
Looking at the symbols.txt generated by `init` reveals the following line:
```
func_020163f4 kind:function(arm,size=0x294) addr:0x20163e0
```
So it seems that `init` now understands that the function starts at 020163f4 and uses a pool at 20163e0. But it incorrectly sets the address of the symbol to the pool's address instead of the function's address. If I manually edit symbols.txt to fix the address, then delink runs without errors for OoE.

---

PoR has the same issue as OoE:
```
>dsd delink --config-path ./config/ACBE00/arm9/config.yaml 
[2024-11-12T22:15:01Z INFO  ds_rom::rom::rom] Loading ROM from ./config/ACBE00/arm9\../../../orig/ACBE00/config.yaml
Error: function func_020a3634 could not be analyzed: InvalidStart
```
```
func_020a3634 kind:function(arm,size=0x294) addr:0x20a3620
```

---

DoS still errors on `init`. The error is different, saying it can't find a function at 02075f18, but there should be a function there.

It seems what's happening is that dsd 0.2.1 detects the `text_end` as being 2067A8C, which is even earlier than 0.2.0 detected it as being (20728f4). (0.2.1 no longer prints the warning about `text_end` being incorrect like 0.2.0, I found that value with a debugger.)

I'm not sure why, as there is definitely a function (not a pool) at 2067A8C:
![image](https://github.com/user-attachments/assets/cc1c8eab-b310-4d21-8b37-63f565859110)

---

Let me know if there's an other info I can provide.

**Comment by LagoLunatic** [comment]
Actually, I'm going to close this issue since the original data-only overlay issue is fixed as of 0.2.1. It'll be too confusing having a bunch of unrelated bugs in one issue so I'll make separate reports with more info.

**Comment by AetiasHax** [comment]
I should've tested everything first before claiming that OoE was compatible... sorry about that. Thanks for providing extra context for the other games, I'll be checking those out soon!
